rdr_service/services/consent/validation.py

```diff
@@ -1,10 +1,171 @@
+        """Find all the expected consents since the minimum date and check the files that have been uploaded"""
+        validation_results = []
+        validated_participant_ids = []
+        for summary in self.consent_dao.get_participants_with_consents_in_range(start_date=min_consent_date):
+            validated_participant_ids.append(summary.participantId)
+            validator = self._build_validator(summary)
+
+            if self._has_new_consent(
+                consent_status=summary.consentForStudyEnrollment,
+                authored=summary.consentForStudyEnrollmentFirstYesAuthored,
+                min_authored=min_consent_date
+            ):
+                validation_results.extend(self._process_validation_results(validator.get_primary_validation_results()))
+            if self._has_new_consent(
+                consent_status=summary.consentForCABoR,
+                authored=summary.consentForCABoRAuthored,
+                min_authored=min_consent_date
+            ):
+                validation_results.extend(self._process_validation_results(validator.get_cabor_validation_results()))
+            if self._has_new_consent(
+                consent_status=summary.consentForElectronicHealthRecords,
+                authored=summary.consentForElectronicHealthRecordsAuthored,
+                min_authored=min_consent_date
+            ):
+                validation_results.extend(self._process_validation_results(validator.get_ehr_validation_results()))
+            if self._has_new_consent(
+                consent_status=summary.consentForGenomicsROR,
+                authored=summary.consentForGenomicsRORAuthored,
+                min_authored=min_consent_date
+            ):
+                validation_results.extend(self._process_validation_results(validator.get_gror_validation_results()))
+
+        results_to_store = []
+        previous_results = self.consent_dao.get_validation_results_for_participants(
+            participant_ids=validated_participant_ids
+        )
+        for possible_new_result in validation_results:
+            if not any(
+                [possible_new_result.file_path == previous_result.file_path for previous_result in previous_results]
+            ):
+                results_to_store.append(possible_new_result)
+
+        self.consent_dao.batch_update_consent_files(results_to_store)
+
+    @classmethod
+    def _process_validation_results(cls, results: List[ParsingResult]):
+        ready_file = cls._find_file_ready_for_sync(results)
+        if ready_file:
+            return [ready_file]
+        else:
+            return results
+
+    @classmethod
+    def _has_new_consent(cls, consent_status, authored, min_authored):
+        return consent_status == QuestionnaireStatus.SUBMITTED and authored > min_authored
+
+    def _build_validator(self, participant_summary: ParticipantSummary) -> 'ConsentValidator':
+        consent_factory = files.ConsentFileAbstractFactory.get_file_factory(
```

Comment:
Should this type hint be an object instead of a string 'ConsentValidator' >

Reply:
Normally they'd just be direct references. But in this case the interpreter wouldn't know about the `ConsentValidator` class yet because it's defined later in the same file. So a string can be used as a placeholder which is evaluated later. https://www.python.org/dev/peps/pep-0484/#forward-references

Reply:
Thanks for the explanation, good to know!